### PR TITLE
chore(model): Make deduplicateTags linear instead of O(n^2) in tag count

### LINKED ISSUE #4109: [chore]: Make deduplicateTags linear instead of O(n^2) in tag count
### Requirement

`deduplicateTags` in `packages/jaeger-ui/src/model/transform-trace-data.ts` removes repeated tags from a span using `Array.prototype.some` inside a `reduce`. For each tag it scans the whole accumulated list, making it O(n^2) in tag count per span.

It runs once per span on every trace load. For normal spans this is fine, but it gets expensive when a span has a lot of tags. GenAI spans are a good example since a long conversation or a big tool schema can expand into many gen_ai.* attributes on a single span.

The fix is to replace the scan with a Set based lookup so it runs in O(n), without changing any existing behavior.

### PR DESCRIPTION
**Pull request by iammdzaidalam**
## Which problem is this PR solving?
- Resolves #4109

## Description of the changes
- Rewrote the tag loop to use a Map of Sets instead of scanning an array
- This improves the time complexity from quadratic to linear
- Made the warning key collision safe by stringifying the key and value
- Fixed mixed-type collisions (e.g., 1 vs "1") in the warnings map
- Made the warning message text completely unambiguous

## How was this change tested?
- Wrote unit tests to verify the duplicate tracking logic
- Added a test to ensure keys with colons do not collide
- Added a test for mixed-type tag collisions
- Ran the local tests and linters

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully: `make lint test`

## AI Usage in this PR (choose one)
See [AI Usage Policy](https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md#ai-usage-policy).
- [x] **Light**: AI provided minor assistance (formatting, simple suggestions)

### CODE REVIEW

File: packages/jaeger-ui/src/model/transform-trace-data.ts [hunk 1]
```diff
@@ -13,14 +13,24 @@ import OtelTraceFacade from './OtelTraceFacade';
...
+  // (rather than a combined key+value string) avoids copying large tag values.
+  const seenValuesByKey = new Map<string, Set<KeyValuePair['value']>>();
+  for (const tag of spanTags) {
+    let seenValues = seenValuesByKey.get(tag.key);
+    if (!seenValues) {
+      seenValues = new Set();
+      seenValuesByKey.set(tag.key, seenValues);
+    }
+    if (seenValues.has(tag.value)) {
+      warningsHash.set(JSON.stringify([tag.key, tag.value]), `Duplicate tag "${tag.key}:${tag.value}"`);
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.ts]: I replaced JSON stringify with a null byte string separator to prevent crashes on BigInt and false collisions on undefined values. As for the Set initialization, TypeScript already correctly infers the exact type from the Map lookup. Adding an explicit type there is completely unnecessary and was left out.

File: packages/jaeger-ui/src/model/transform-trace-data.test.js [hunk 2]
```diff
@@ -41,6 +41,43 @@ describe('deduplicateTags()', () => {
...
+      { key: 'a', value: 'b:c' },
+      { key: 'a', value: 'b:c' },
+    ]);
+
+    expect(tagsInfo.tags).toEqual([
+      { key: 'a:b', value: 'c' },
+      { key: 'a', value: 'b:c' },
+    ]);
+    // Both duplicated pairs are tracked separately; neither warning is dropped.
+    expect(tagsInfo.warnings).toHaveLength(2);
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.test.js]: updated the test to assert the exact array of strings instead of just the length.

File: packages/jaeger-ui/src/model/transform-trace-data.ts [hunk 3]
```diff
@@ -13,14 +13,24 @@ import OtelTraceFacade from './OtelTraceFacade';
 // exported for tests
 export function deduplicateTags(spanTags: ReadonlyArray<KeyValuePair>) {
   const warningsHash: Map<string, string> = new Map<string, string>();
-  const tags: KeyValuePair[] = spanTags.reduce<KeyValuePair[]>((uniqueTags, tag) => {
-    if (!uniqueTags.some(t => t.key === tag.key && t.value === tag.value)) {
-      uniqueTags.push(tag);
+  const tags: KeyValuePair[] = [];
+  // Track seen values per key so duplicate detection is a single O(n) pass
+  // instead of rescanning the accumulated list for every tag. Keying by value
+  // (rather than a combined key+value string) avoids copying large tag values.
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.ts]: comment is correct btw... the string copy only happens on the error path, so performance isn't affected.

File: packages/jaeger-ui/src/model/transform-trace-data.ts [hunk 4]
```diff
@@ -13,14 +13,24 @@ import OtelTraceFacade from './OtelTraceFacade';
...
+  const seenValuesByKey = new Map<string, Set<KeyValuePair['value']>>();
+  for (const tag of spanTags) {
+    let seenValues = seenValuesByKey.get(tag.key);
+    if (!seenValues) {
+      seenValues = new Set();
+      seenValuesByKey.set(tag.key, seenValues);
+    }
+    if (seenValues.has(tag.value)) {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/model/transform-trace-data.ts]: redundant check if above if() was successful.
```
  for (const tag of spanTags) {
    const values = seen.get(tag.key);
    if (!values || !values.has(tag.value)) {
      if (values) {
        values.add(tag.value);
      } else {
        seen.set(tag.key, new Set([tag.value]));
      }
      tags.push(tag);
    } else {
      warningsHash.set(`${tag.key}:${tag.value}`, `Duplicate tag "${tag.key}:${tag.value}"`);
    }
  }
```

Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.ts]: thanks, will update it... just one thing, thinking of keeping  \0  for the warnings key tho, just to be safe if a key or value has a colon

File: packages/jaeger-ui/src/model/transform-trace-data.ts [hunk 5]
```diff
@@ -13,14 +13,21 @@ import OtelTraceFacade from './OtelTraceFacade';
...
+      if (values) {
+        values.add(tag.value);
+      } else {
+        seen.set(tag.key, new Set([tag.value]));
+      }
+      tags.push(tag);
     } else {
-      warningsHash.set(`${tag.key}:${tag.value}`, `Duplicate tag "${tag.key}:${tag.value}"`);
+      warningsHash.set(`${tag.key}\0${String(tag.value)}`, `Duplicate tag "${tag.key}:${tag.value}"`);
     }
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.ts:32]: imo it looks like the diff formatting mixed up the old and new lines here though. I'm going to stick with the original string casting behavior so we don't expand the scope of this perf fix.

[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.ts:32]: well since I already touched the colon collisions I figured I might as well fix the type collision too to make it bulletproof. fixed all the above, added the tests, and updated the PR description.

File: packages/jaeger-ui/src/model/transform-trace-data.ts [hunk 6]
```diff
@@ -13,14 +13,21 @@ import OtelTraceFacade from './OtelTraceFacade';
...
+    if (!values || !values.has(tag.value)) {
+      if (values) {
+        values.add(tag.value);
+      } else {
+        seen.set(tag.key, new Set([tag.value]));
+      }
+      tags.push(tag);
     } else {
-      warningsHash.set(`${tag.key}:${tag.value}`, `Duplicate tag "${tag.key}:${tag.value}"`);
+      warningsHash.set(`${tag.key}\0${String(tag.value)}`, `Duplicate tag "${tag.key}:${tag.value}"`);
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.ts]: updated the warning format to explicitly state the key and value

File: packages/jaeger-ui/src/model/transform-trace-data.test.js [hunk 7]
```diff
@@ -41,6 +41,43 @@ describe('deduplicateTags()', () => {
     ]);
     expect(tagsInfo.warnings).toEqual(['Duplicate tag "b.ip:8.8.4.4"']);
   });
+
+  it('collapses repeated duplicates into a single warning and keeps the first', () => {
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.test.js:45]: added the test cases to cover the type collision edge cases

File: packages/jaeger-ui/src/model/transform-trace-data.test.js [hunk 8]
```diff
@@ -41,6 +41,43 @@ describe('deduplicateTags()', () => {
...
+      { key: 'x', value: 'a' },
+      { key: 'x', value: 'a' },
+      { key: 'x', value: 'a' },
+    ]);
+
+    expect(tagsInfo.tags).toEqual([{ key: 'x', value: 'a' }]);
+    expect(tagsInfo.warnings).toEqual(['Duplicate tag "x:a"']);
+  });
+
+  it('does not collide when key or value contains a colon', () => {
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.test.js:56]: done

File: packages/jaeger-ui/src/model/transform-trace-data.test.js [hunk 9]
```diff
@@ -41,6 +41,43 @@ describe('deduplicateTags()', () => {
...
+    const tagsInfo = deduplicateTags([
+      { key: 'a:b', value: 'c' },
+      { key: 'a', value: 'b:c' },
+    ]);
+
+    expect(tagsInfo.tags).toHaveLength(2);
+    expect(tagsInfo.warnings).toEqual([]);
+  });
+
+  it('reports both duplicates when colliding key/value pairs are each duplicated', () => {
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply packages/jaeger-ui/src/model/transform-trace-data.test.js:66]: added

**Review by yurishkuro — APPROVED** [review]
